### DISCUSSION TITLE
New Plugin - AIO controll via liquidctl

### DISCUSSION THREAD

**Discussion by jmarucha** [Rem0o/FanControl.Releases#797] [Show and tell]
If you ever wanted to get the data from temperature probe your shiny AIO has for liquid (NZXT Krakens do for example), and bind it to fan curves, yeah, you're in the right place.

### Just give me it already

Release: ~FanControl.Liquidctl v0.1.0~ [FanControl.Liquidctl v0.1.1](https://github.com/jmarucha/FanControl.Liquidctl/releases/tag/v0.1.1)
Repo: [jmarucha/FanControl.Liquidctl](https://github.com/jmarucha/FanControl.Liquidctl)

### Features

- ~Support for AIOs from extensive list of liquidctl [supported devices](https://github.com/liquidctl/liquidctl#supported-devices) (disclaimer: only one model tested so far ;) )~ No, it does not. So far Asetec 690LC devices are supported (NZXT Kraken X0&X1, non-iCUE Corsairs hydro and EVGL CLCs), and other Krakens of X series. Maybe some others. I am currently in process of talking with liquidctl team to unify the CLI a little bit to make support of other stuff doable.
- Software pump RPM probe and pump control (I found it a bit more reliable than fan header from MB, because if any software does *anything* via USB to NZXT Krakens, header control is not working anymore. And NZXT Cam is a sneaky bastard).
- LIQUID TEMPERATURE PROBE - if you bought AIO, you most likely did it for this massive heat capacity allowing rather smooth fan reactions. And CPU fan headers doesn't care about that, only on CPU temps. Guess what, you can override that.

### How I use it

Yes, I made this plugin mostly to myself, but it would be a big waste not to share it. So, my setup is:
- Rear and top fans exhaust
- Front AIO radiator with intake fans
- Super bulky nVidia GPU
And how I wanted to use it (which resulted in the birth of the plugin):
- AIO pump controlled by CPU temps. Fluid flow is what cools down CPU directly, so I need a feedback loop there.
- Exhaust temps depending on rather general mixture of temps inside. Currently just GPU temp because it's the most power hungry component.
- Radiator/intake basing on mixture of the fluid temperature and long-time average of CPU temp. Long enough to avoid fast ramp-ups, short enough to react to load preemptively.

### How you can use it

However you want. That includes your modifications of the plugin (development setup in README.md).

### Can I contribute?

YES! First, I am looking for people testing if the plugin works on any other machine that mine (I don't have more desktops with AIOs). Bug reports are most welcome in the issues section (however you may find it's rather a bug in liquidctl, then report it there).

Suggestions and PRs are welcome as well.

### I want to see.

Here we go:

![Pump control](https://github.com/jmarucha/FanControl.Liquidctl/raw/master/docs/images/PumpControl.png)
![Temperature sensor](https://github.com/jmarucha/FanControl.Liquidctl/raw/master/docs/images/FluidTemp.png)

**Comment by Fr0stX76** [comment]
Nice! Ive been using  liquidctl to control my corsair PSU fan via batch files thru HWinfo triggers. I will test this tonight as if this work it will be so much nicer and efficient to control it via graphs!

For reference, people get the idea that any PSU can fan can be controlled by this, I have an HX850i which is part of the supported devices of liquidctl.

**Reply by jmarucha** [reply]
Can you get me the output for "liquidctl status --json"?
I think I didn't add fan channels and I kinda not know how they identify

EDIT: I'm sure I didn't add fan channels. I get the output from the status by looking for keys "Pump duty" "Pump speed" and "Liquid temperature" and judging by liquidctl readme there's stuff like "Fan speed" and "Fan duty". Will dig into that later (tho annoyingly I have no way of testing it).

**Reply by jmarucha** [reply]
*actually not needed. I will figure that out from the code. Anyway I plan to add support to different devices, as now it shall only work fror NZXT Krakens (all of them), and some EVGA and Corsair AIOs

**Comment by reelmccoy** [comment]
Been using LiquidCTL for a while now and gave this a shot on my system with an NZXT Kraken x72 with a "NZXT Smart Device V2" in the mix as Device #1 in LiquidCTL. Running into an exception on ReadStatus when I'm guessing NewtonSoft is trying to handle a fan control mode with a value of null. I can feed my status JSON and exception info if needed.

**Comment by Rem0o** [comment]
When this gets polished up a bit, will include it in the fancontrol README.

**Comment by jmarucha** [comment]
Please do!

Sent from Nine
________________________________

**Reply by reelmccoy** [reply]
Here's a sample output from LiquidCTL's status.

`[{"bus": "hid", "address": "\\\\?\\hid#vid_1e71&pid_170e#a&f2d37d0&2&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}", "description": "NZXT Kraken X (X42, X52, X62 or X72)", "status": [{"key": "Liquid temperature", "value": 39.1, "unit": "°C"}, {"key": "Fan speed", "value": 1948, "unit": "rpm"}, {"key": "Pump speed", "value": 2934, "unit": "rpm"}]}, {"bus": "hid", "address": "\\\\?\\hid#vid_1e71&pid_2006#9&2d827572&2&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}", "description": "NZXT Smart Device V2", "status": [{"key": "Fan 1 control mode", "value": null, "unit": ""}, {"key": "Fan 1 duty", "value": 50, "unit": "%"}, {"key": "Fan 1 speed", "value": 0, "unit": "rpm"}, {"key": "Fan 2 control mode", "value": null, "unit": ""}, {"key": "Fan 2 duty", "value": 50, "unit": "%"}, {"key": "Fan 2 speed", "value": 0, "unit": "rpm"}, {"key": "Fan 3 control mode", "value": null, "unit": ""}, {"key": "Fan 3 duty", "value": 50, "unit": "%"}, {"key": "Fan 3 speed", "value": 0, "unit": "rpm"}, {"key": "Noise level", "value": 96, "unit": "dB"}]}]`

I gave an attempt at adding removing null checking for the value but then face a different exception when the plugin gets added that I think is more in FanControl and doing a little reflection? Possibly a build thing in my own environment. I know enough to be dangerous but not enough C#/.NET to be effective. lol

**Reply by reelmccoy** [reply]
Just realizing that 2nd bit of data is likely coming from an NZXT fan controller but no fans are connected. This was a NZXT LetsBld system with an H710i case that comes with the fan controller.

**Reply by jmarucha** [reply]
https://github.com/jmarucha/FanControl.Liquidctl/releases/tag/v0.1.1
Hey, I fixed it. Indeed problem was with nulls. I hope it works now!

**Reply by jmarucha** [reply]
@reelmccoy does it work now?

**Comment by reelmccoy** [comment]
Sorry for not getting back sooner. Been battling my own programming nightmare. lol. 

Anyway, I still run into an issue. The log mentions an unhandled exception in FanControl but with no stack trace so no clue where the issue is. The next log entry is "Autofac.Core.DependencyResolutionException: An exception was thrown while activating FanControl.ViewModels.MainWindowViewModel.". I'll attach the log text file on my last attempt. 

I can also start moving discussion over to the project's github if desired.

[log.txt](https://github.com/Rem0o/FanControl.Releases/files/8756225/log.txt)

**Reply by Rem0o** [reply]
Seems like there is a null ref here:
```
FanControl.Domain.BackendProviders.Plugin.PluginSensor.get_Name()
```

Meaning one of the sensors name getter has a null ref, or there is a "null" sensor in the IPluginSensorsContainer.

**Comment by reelmccoy** [comment]
I did take a quick look at the potential LiquidCTL commands to see if perhaps specific commands might be tripping things up but I can't get LiquidCTL happy with --address requests using the far out response I get from status. I even tried slipping in "-m kraken" instead of using the address but I was likely looking at the wrong thing.

I may try another look at this later tonight if jmarucha doesn't beat me to it.

**Comment by reelmccoy** [comment]
May have found the culprit. Something that shows up with my Kracken X72 since it doesn't have Pump Duty. With the code tweak, I now have the temperature information and pump speed showing in FanControl. Can't set the fan speed or pump speed but hey! Progress!

The issue is in the constructor in LiquidctlDevice.cs. hasPumpSpeed and hasPumpDuty are swapped as far as being set and then checked. Created an issue over on the project's GitHub for now. And since I know what the fix is, I can try making the change myself but I'm a total GitHub noob. Haven't ever made any change requests with GitHub projects.

**Comment by reelmccoy** [comment]
I should probably figure the contribution part as I've got fan speed of the X72 reporting added as well. All that's missing is setting the pump and fan percentages as neither is being reported with the X72.

**Reply by reelmccoy** [reply]
Throw in setting the fan speed and pump speed on the X72 in quick tests. Added control sensors only if "Fan Speed" reported with LiquidCtl. If they aren't set to a curve, they get reset to 100%.

**Comment by reelmccoy** [comment]
Yup. I should figure out how contribute. Have the ability to set the fan speed and pump speed (if pump duty isn't in the status json) which appears to work with my Kraken x72. I did it that way as Kraken z73s may also return pump duty and fan speed but I'm not sure.

**Comment by sysyphus** [comment]
Is there any way to make this plugin works with Evga CLC?
